test/errtest.c

```diff
@@ -36,25 +36,40 @@ static int test_print_error_format(void)
     char *out = NULL, *p = NULL;
     int ret = 0, len;
     BIO *bio = NULL;
+    int syserr = EPERM;
+    int reasoncode;
 
-    BIO_snprintf(expected, sizeof(expected), expected_format, OPENSSL_FUNC);
+    ERR_PUT_error(ERR_LIB_SYS, 0, syserr, "errtest.c", 30);
+    reasoncode = ERR_GET_REASON(ERR_peek_error());
 
-    if (!TEST_ptr(bio = BIO_new(BIO_s_mem())))
-        return 0;
+    if (reasoncode != syserr) {
+        TEST_skip("libcrypto didn't register EPERM (%d) correctly (%d).  "
+                  "This is a known limitation",
+                  syserr, reasoncode);
+        ERR_clear_error();
+    } else {
+        char buf[256];
+
+        OPENSSL_strlcpy(buf, strerror(syserr), sizeof(buf));
+        BIO_snprintf(expected, sizeof(expected), expected_format,
+                     OPENSSL_FUNC, buf);
 
-    ERR_PUT_error(ERR_LIB_SYS, 0, 1, "errtest.c", 30);
-    ERR_print_errors(bio);
+        if (!TEST_ptr(bio = BIO_new(BIO_s_mem())))
+            return 0;
 
-    if (!TEST_int_gt(len = BIO_get_mem_data(bio, &out), 0))
-        goto err;
-    /* Skip over the variable thread id at the start of the string */
-    for (p = out; *p != ':' && *p != 0; ++p) {
-        if (!TEST_true(IS_HEX(*p)))
+        ERR_print_errors(bio);
+
+        if (!TEST_int_gt(len = BIO_get_mem_data(bio, &out), 0))
+            goto err;
+        /* Skip over the variable thread id at the start of the string */
+        for (p = out; *p != ':' && *p != 0; ++p) {
+            if (!TEST_true(IS_HEX(*p)))
+                goto err;
+        }
+        if (!TEST_true(*p != 0)
+            || !TEST_strn_eq(expected, p, strlen(expected)))
```

Comment:
I suggest pointing out in a comment that the string is compared only up to the reason string (and subsequent `:`).

Reply:
Uhmmm, but that's not true

Reply:
This is only comparing those strings up to `strlen(expected)`, and the one in `p` is longer.